File: pieces/BishopBehaviour.py
from typing import List

from Array import Array2D, get_diagonal_positions, sort_by_distance
from custom_types import Piece, Position


# @profile
def allowed_moves(
    piece: Piece, board: Array2D, pos: Position, is_white: bool
) -> List[Position]:
    diagonal_positions = get_diagonal_positions(pos, board.shape)
    filt_all_ms = []
    for positions in diagonal_positions:
        for dir_positions in positions:
            dir_positions = sort_by_distance(piece.position, dir_positions)
            for move_pos in dir_positions:
                mp_piece = board[move_pos[0]][move_pos[1]]
                if isinstance(mp_piece, Piece):
                    break
                else:
                    filt_all_ms.append(move_pos)
    return filt_all_ms


# @profile
def allowed_takes(
    piece: Piece, board: Array2D, pos: Position, is_white: bool
) -> List[Position]:
    diagonal_positions = get_diagonal_positions(pos, board.shape)
    filt_all_ts = []
    for positions in diagonal_positions:
        for dir_positions in positions:
            dir_positions = sort_by_distance(piece.position, dir_positions)
            for take_pos in dir_positions:
                tp_piece = board[take_pos[0]][take_pos[1]]
                if isinstance(tp_piece, Piece):
                    if tp_piece.is_white != piece.is_white:
                        filt_all_ts.append(take_pos)
                    break
    return filt_all_ts

File: src/pieces/BishopBehaviour.rs
use crate::{
    types::{Array2D, Position, PositionContent, Shape},
    Array::{get_diagonal_positions, sort_by_distance},
};

use super::piece::Piece;

pub(crate) fn allowed_moves(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let diagonal_positions = get_diagonal_positions(pos, shape);
    let mut filt_all_ms: Vec<Position> = Vec::with_capacity(diagonal_positions.len());
    for positions in diagonal_positions.into_iter() {
        for dir_positions in positions {
            let dir_positions = sort_by_distance(pos, dir_positions);
            for move_pos in dir_positions.into_iter() {
                let mp_piece: &PositionContent = &board[usize::try_from(move_pos.0).unwrap()]
                    [usize::try_from(move_pos.1).unwrap()];
                match mp_piece {
                    PositionContent::PieceContent(_) => filt_all_ms.push(move_pos),
                    PositionContent::Empty => break,
                }
            }
        }
    }
    filt_all_ms
}

pub(crate) fn allowed_takes(
    piece: &Piece,
    board: &Array2D,
    pos: &Position,
    shape: &Shape,
    is_white: bool,
) -> Vec<Position> {
    let diagonal_positions = get_diagonal_positions(pos, shape);
    let mut filt_all_ts: Vec<Position> = Vec::with_capacity(diagonal_positions.len());
    for positions in diagonal_positions.into_iter() {
        for dir_positions in positions {
            let dir_positions = sort_by_distance(pos, dir_positions);
            for take_pos in dir_positions.into_iter() {
                let tp_piece: &PositionContent = &board[usize::try_from(take_pos.0).unwrap()]
                    [usize::try_from(take_pos.1).unwrap()];
                match tp_piece {
                    PositionContent::PieceContent(tp_piece) => {
                        if tp_piece.is_white != piece.is_white {
                            filt_all_ts.push(take_pos)
                        }
                        break
                    },
                    PositionContent::Empty => (),
                }
            }
        }
    }
    filt_all_ts
}
